### ISSUE TITLE
Ambiguous conversion of Python lists into either Java arrays or Lists

### ISSUE THREAD

**Issue by gciatto-unibo** [jpype-project/jpype#998]
I noticed that, if a java method has two overloads accepting an array and an iterable, respectively, passing a Python list (or any other sort of iterable) to that method produces an exception.

Consider for instance the class:
```java
class Foo {
  Y method(X[] args)
  Y method(Iterable<X> args) 
} 
```
this can be used from Python as follows:
```python
foo = Foo() 
foo.method([a1, a2, a3,...]) 
```

I would expect the second overload to be called in this case (meaning that the Python list is converted into a Java iterable by Jpype), yet an exception is raised instead.

Is this intended?

IMHO whenever the actual argument is a Python iterable, and an overload exists accepting an iterable, that overload should be preferred.

**Comment by Thrameos** [comment]
As described in the documentation the system used 4 levels to define the quality of a match.  These are exact, implicit, explicit and none.   In this case, a Python list can either be converted to a list or a iterable.   Both conversions are possible and thus candidates.   As both require conversion, they are both at the same level (``implicit``) and thus ambiguous so an exception is thrown.   The solution to ambiguity is to apply a cast to select which one the user intends.

```
foo.method(Iterable@[a1, a2, a3,...])
```
or
```
foo.method(X[:]@[a1,a2,a3,...])
```
 
Casting should resolve your issue.   But beyond that lets consider if it would be technically possible to have a preference.

There are some issues  to consider here.   We could add additional levels (exact, preferred, implicit, explicit, none) that allow one to have preference over another, but this is difficult because there is no way to know what types of overloads one may have and thus you may just end up with the difficulty that you have two "preferred" overloads competing against each other.   The second is mean you have multiple arguments.   Consider the case of 3 arguments.  Is it better to match "preferred, exact, implicit" or "implicit, exact, exact"?  Basically this leads to endless corner cases.   For that reason that change has been rejected previously.

There is another way that we can give preference without changing the current leveling scheme and that is the method resolution.   When the list of methods is enumerated the first time, one overload can be marked to hide another.   If two are matching and the first hides the other, it will resolve ambiguity by disregarding the second.   This is done by constructing a sorted list of resolutions with the second marked with the id of the first.   When resolving overloads they are scanned in order, the first will match, and the second will match but it will then look at the hides list to find that it was already covered by the first and disregard as a candidate.  The code to determine which methods hide others is written in Java and is nominally loaded with the Java preference rules.   There is one current bug in that it does not recognize the case in which a specific overload should mask a variadic argument list (``foo(Object, Object)`` should hide ``foo(Object...)``).   But as Java does not have the concept of conversions, there are no rules regarding comparing a collection to an array.

We could add one that object types hide raw arrays.   This would not break existing code because it would currently be ambiguous (an error) and thus no existing working code will break.  I would be somewhat hesitant to impose such a rule as it is difficult to see all the edge cases that would be generated.   Should it just be Iterable that  covers arrays?   How about Collection?  What about a user defined collection class or stream?  Would adding this rule cause conflicts when we allow Python objects to map directly to Java types  (PyList type)?

There is one other technical issue to consider which is the cache.   When repeated instances of a method are called with the same argument type  it bypasses the resolution method entirely and just checks if the previous overload still works, if it does it uses it without further interregation.   This could be problematic if there is a preference issue.   Suppose we are matching to a python list against ``Iterable<>`` and ``String[]``.  If I pass in a python list of strings with a cast I hit ``String[]`` on the first call, but then on the next call I send in a python list of ints.   Both types have the cache hash id of ``python list``, so the second call consults the cache and fails (int!=String) so it goes back to resolution which would find ``Iterable<>``.   Great!  But now if I flip the order of operations, the first call goes to ``Iterable<>`` as a list of ints is convertible.   The second call has the same cache id as it also is python list cache id.   But this time ``Iterable<>`` is also a match so it uses it immediately without going to resolution.   So you get different behaviors depending on different orders.   Forcing the user to cast each time prevents the cache system from failing because it requires that we can't chose preference based on the contents of a list.  

Sorry for the long winded technical details here.   The bottom line is that I understand the preference that you have, but there are many technical details that are driving it towards declaring it ambiguous in the current implementation.   It is not insurmountable, but we would need a compelling demonstration that it could be resolved regardless of the order of operations before we can accept a PR to force a preference.

**Comment by gciatto-unibo** [comment]
Oh thank you for the very nice explanation.
I totally understand.

Furthermore, I wasn't aware of the `Type@object` syntax: is that documented somewhere?
Thank you again.

**Comment by Thrameos** [comment]
It is mentioned in the quickstart guide, but there appears to be a formatting issue such that the text of the example does not appear in the web version.  Not sure why it is missing though the footnote appears be correct.

The guide should read

```
+---------------------------+---------------------------------------------------------+---------------------------------------------------------+
|                           |                                                         |                                                         |
| Casting to a specific     | .. code-block:: java                                    | .. code-block:: python                                  |                                   | type [11]_                |                                                         |                                                         |
|                           |     BaseClass b = (BaseClass)myObject;                  |     b = (BaseClass) @ myObject                          |
|                           |                                                         |                                                         |
+---------------------------+---------------------------------------------------------+---------------------------------------------------------+
```

The missing Java casting operator syntax ``(Type)object)`` is written as Python ``Type@object``

There is also an example regarding the use of Functional operators that appears to suffer from the same formatting issue.

```
u=DoubleUnaryOperator@(lambda x:x*2)
```

Please be so kind as reviewing #999 to see if it addresses the documentation requirements.
